### ISSUE TITLE
[nanodet-plus] Conv layer shape wrong

### ISSUE THREAD

**Issue by vuthithao** [PINTO0309/onnx2tf#15]
### Issue Type

Others

### onnx2tf version number

1.1.22

### Download URL for ONNX

https://drive.google.com/file/d/1HyUDTYSTgS7_Gs29SxUSoxs3irqvIsXM/view?usp=sharing

### Parameter Replacement JSON

```json
{}
```

### Description

1. Purpose: Personal development
2. What:
Running script: `onnx2tf -i nanodet-plus-m_416.onnx -oh5 -k "data"`
Issue log:
```
INFO: onnx_op_type: Conv onnx_op_name: /backbone/stage2/stage2.1/branch2/branch2.0/Conv
INFO: input_name.1: /backbone/stage2/stage2.1/Split_output_1 shape: [1, 58, 52, 52] dtype: float32
INFO: input_name.2: onnx::Conv_1392 shape: [58, 58, 1, 1] dtype: <class 'numpy.float32'>
INFO: input_name.3: onnx::Conv_1393 shape: [58] dtype: <class 'numpy.float32'>
INFO: output_name.1: /backbone/stage2/stage2.1/branch2/branch2.0/Conv_output_0 shape: [1, 58, 52, 52] dtype: float32
ERROR: The trace log is below.
Traceback (most recent call last):
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/onnx2tf/utils/common_functions.py", line 261, in print_wrapper_func
    result = func(*args, **kwargs)
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/onnx2tf/utils/common_functions.py", line 323, in inverted_operation_enable_disable_wrapper_func
    result = func(*args, **kwargs)
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/onnx2tf/ops/Conv.py", line 153, in make_node
    tf.nn.convolution(
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/keras/layers/core/tf_op_layer.py", line 119, in handle
    return TFOpLambda(op)(*args, **kwargs)
  File "/mnt/hdd10tb/Users/thaovu/.conda/envs/onnx2tf/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
ValueError: Exception encountered when calling layer "tf.nn.convolution_4" (type TFOpLambda).

Depth of input (52) is not a multiple of input depth of filter (58) for '{{node tf.nn.convolution_4/convolution}} = Conv2D[T=DT_FLOAT, data_format="NHWC", dilations=[1, 1, 1, 1], explicit_paddings=[], padding="SAME", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true](Placeholder, tf.nn.convolution_4/convolution_internal/filters)' with input shapes: [1,58,52,52], [1,1,58,58].
```

Question 1: Shape of onnx::Conv_1393 layer must be [58,58,1,1], but in here it is [1,1,58,58]. Any cause from data_format?
Question 2: I set -k "data" but data_format of this layer is NHWC. Correct me if I wrong
3. How: 
Firstly, I ran without -k option and get the same issue, I tried to use -k but still the the same
I also check parameter replacement but cant find suitable solution.
4. Why: I wanna convert NanoDet-plus-m model to Keras (h5 format)
5. Resources: Pytorch model from NanoDet model zoo https://github.com/RangiLyu/nanodet
Thank you for your hard working. Your projects are very useful <3

**Comment by PINTO0309** [comment]
The most likely problem with this tool is `Reshape` or `Transpose`, as described in the README. While the tool tries to automatically adjust model transformations as much as possible to avoid breakdowns, there are limits to automatic geometry changes and transpositions. The combination of Reshape with a 5-dimensional transposition is particularly error-prone. This applies to most of the shuffle-net models.

Therefore, when a conversion error occurs, it is necessary to check around `Reshape` and `Transpose` to see if the conversion is in a different state than expected. Below is a flowchart of how to identify the problem and how to resolve it.

1. Identify the name of the OP where the error finally occurred.
`/backbone/stage2/stage2.1/branch2/branch2.0/Conv`
![image](https://user-images.githubusercontent.com/33194443/201913708-df90feee-e596-45c4-8b99-21a65168ae73.png)
![image](https://user-images.githubusercontent.com/33194443/201913626-86ddacdd-ebf8-4a70-9a6b-4d5b6736d376.png)
![image](https://user-images.githubusercontent.com/33194443/201914033-9dec0011-76fc-4fe6-878b-82697bf770f6.png)
2. Trace back a short distance from the OP where the error finally occurred to the previous OP to find where the `Reshape` or `Transpose` shape change is broken.
![image](https://user-images.githubusercontent.com/33194443/201914627-7b93fc40-c7fc-4ee4-982c-15de1aed5d70.png)
3. Check the conversion logs for the three OPs in the figure above.
    ```
    INFO: onnx_op_type: Reshape onnx_op_name: /backbone/stage2/stage2.0/Reshape
    INFO: input_name.1: /backbone/stage2/stage2.0/Concat_output_0 shape: [1, 116, 52, 52] dtype: float32
    INFO: input_name.2: /backbone/stage2/stage2.0/Constant_output_0 shape: [5] dtype: <class 'numpy.int64'>
    INFO: output_name.1: /backbone/stage2/stage2.0/Reshape_output_0 shape: [1, 2, 58, 52, 52] dtype: float32
    INFO: tf_op_type: reshape
    INFO: input.1.tensor: name: tf.compat.v1.transpose/transpose:0 shape: (1, 116, 52, 52) dtype: <dtype: 'float32'> 
    INFO: input.2.shape: val: [1, 2, 58, 52, 52] 
    INFO: output.1.output: name: tf.reshape/Reshape:0 shape: (1, 2, 58, 52, 52) dtype: <dtype: 'float32'> 
    
    INFO: onnx_op_type: Transpose onnx_op_name: /backbone/stage2/stage2.0/Transpose
    INFO: input_name.1: /backbone/stage2/stage2.0/Reshape_output_0 shape: [1, 2, 58, 52, 52] dtype: float32
    INFO: output_name.1: /backbone/stage2/stage2.0/Transpose_output_0 shape: [1, 58, 2, 52, 52] dtype: float32
    INFO: tf_op_type: transpose_v2
    INFO: input.1.a: name: tf.reshape/Reshape:0 shape: (1, 2, 58, 52, 52) dtype: <dtype: 'float32'> 
    INFO: input.2.perm: val: [0, 2, 1, 3, 4] 
    INFO: output.1.output: name: tf.compat.v1.transpose_1/transpose:0 shape: (1, 58, 2, 52, 52) dtype: <dtype: 'float32'> 
    
    INFO: onnx_op_type: Reshape onnx_op_name: /backbone/stage2/stage2.0/Reshape_1
    INFO: input_name.1: /backbone/stage2/stage2.0/Transpose_output_0 shape: [1, 58, 2, 52, 52] dtype: float32
    INFO: input_name.2: /backbone/stage2/stage2.0/Constant_1_output_0 shape: [4] dtype: <class 'numpy.int64'>
    INFO: output_name.1: /backbone/stage2/stage2.0/Reshape_1_output_0 shape: [1, 116, 52, 52] dtype: float32
    INFO: tf_op_type: reshape
    INFO: input.1.tensor: name: tf.compat.v1.transpose_2/transpose:0 shape: (1, 58, 2, 52, 52) dtype: <dtype: 'float32'> 
    INFO: input.2.shape: val: [1, -1, 52, 52] 
    INFO: output.1.output: name: tf.reshape_1/Reshape:0 shape: (1, 116, 52, 52) dtype: <dtype: 'float32'> 
    ```
    |OP|matters for investigation|
    |:-|:-|
    |/backbone/stage2/stage2.0/Reshape|Correct because the output shape is (1, 2, 58, 52, 52)|
    |/backbone/stage2/stage2.0/Transpose|(1, 58, 2, 52, 52), which is an incorrect transposition. The correct value should be (1,58,2,52,52).|
4. We found an error in the transpose behavior of `/backbone/stage2/stage2.0/Transpose`, so we will change the behavior of `Transpose`. To do so, create a JSON file to rewrite Transpose's perm as explained in the [README - Parameter replacement](https://github.com/PINTO0309/onnx2tf#parameter-replacement).

    The perm for transposing from the input shape (1, 2, 58, 52, 52) to the output shape (1, 58, 2, 52, 52) is [0,2,1,3,4]. Therefore, rewrite the perm of `Transpose` by describing the JSON file as follows. I wrote "rewrite," but we are actually just creating a JSON file here.
    - replace.json
        ```json
        {
          "format_version": 1,
          "operations": [
            {
              "op_name": "/backbone/stage2/stage2.0/Transpose",
              "param_target": "attributes",
              "param_name": "perm",
              "values": [0,2,1,3,4]
            },
            {
              "op_name": "/backbone/stage2/stage2.0/Reshape_1",
              "param_target": "outputs",
              "param_name": "/backbone/stage2/stage2.0/Reshape_1_output_0",
              "post_process_transpose_perm": [0,2,3,1]
            }
          ]
        }
        ```
5. To execute the conversion command again while loading the created JSON file into the tool, enter the command as follows
    ```
    onnx2tf -i nanodet-plus-m_416.onnx -prf replace.json
    ```
6. The error occurs again, but this is because there are several similar Reshape and Transpose combinations. Although an error occurred, we can see that the error occurred in a different OP than the previous one. This means that the first Conv with an error was successfully converted, and so on up to the next Conv.

    `/backbone/stage2/stage2.2/branch2/branch2.0/Conv`
    ![image](https://user-images.githubusercontent.com/33194443/201921582-d4a357a3-5530-43cc-842d-9eefe0154708.png)
7. Thereafter, repeat steps 1. through 5. until there are no more errors. The model you shared seems to have 16 identical Reshape - Traspose - Reshape combinations. Thus, it would be necessary to write similar statements 16 times in the JSON file.

**Comment by vuthithao** [comment]
I tried to replace parameter as you comment with replace.json here https://drive.google.com/drive/folders/1bDJw2mJVGirdpO33POvDxrEQx0cTD2HY?usp=sharing
And get issue at Resize layer
![image](https://user-images.githubusercontent.com/31527427/202084873-fe97f211-77f7-4946-b868-7ad723efff96.png)
With len(graph_node.inputs)=2, this case fall in "else" of https://github.com/PINTO0309/onnx2tf/blob/58b0c6b2e26fbaa732b96338bb8d8d08c1368c8e/onnx2tf/ops/Resize.py#L149
I simply add "else" case:
   ```
 # Tensorflow require the shape of "size" in the "tf.image.resize" must be known at
    # graph creation time. However in the dynamic shape situation, the shape of "new_size"
    # will be "None", the actual shape can only be determine at runtime. But we know
    # "new_size" should always contain [h, w], therefore the shape must be 2.
    else:
        if hasattr(graph_node.outputs[0], 'shape') \
            and graph_node.outputs[0].shape is not None \
            and isinstance(graph_node.outputs[0].shape[-2], int) \
            and isinstance(graph_node.outputs[0].shape[-1], int):
            new_size = graph_node.outputs[0].shape[-2:len(graph_node.outputs[0].shape)] # Estimated from ONNX output shape
    if hasattr(new_size, 'set_shape'):
        new_size.set_shape([2])
```
It works.
Thank for your supports.

**Comment by PINTO0309** [comment]
Thanks for pointing that out. I found the cause of the error in `Resize` OP.
Only the models with opset<=10 seemed to have errors. The number and type of inputs seem to be completely different. Originally, the tool focused on testing models with opset >= 11 and was inadequate for models with opset <= 10.

- opset = 10 https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Resize-10
- opset > 10 https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Resize-11

Therefore, the modifications you suggested would have been fine, but I have made a more generic modification.
commit: https://github.com/PINTO0309/onnx2tf/commit/b2757407e33e5744884188b2e683f689eae8379a
release: https://github.com/PINTO0309/onnx2tf/releases/tag/1.1.23

I have released 1.1.23 and upgrading to the latest package should solve the problem.

**Comment by PINTO0309** [comment]
Fixes: https://github.com/PINTO0309/onnx2tf/releases/tag/1.1.29
